### ISSUE TITLE
Utility Iterable implementation for taylor series coefficients [enhancement]

### ISSUE THREAD

**Issue by axkr** [kredel/java-algebra-system#28]
Can you please add an `Iterable<BigRational>` (or generic parameter?)  implementation for taylor series coefficients calculation of rational polynomials `numerator/denominator` ?

- https://docs.oracle.com/javase/8/docs/api/java/lang/Iterable.html

the constructor should have a signature similar to this:
```
SeriesCoefficientsIterator(GenPolynomial<BigRational> numerator, GenPolynomial<BigRational> denominator, String var, BigRational varExpansionValue, int order)
```

The `next()` method of the iterator should create the coefficients up to `order`.

**Comment by kredel** [comment]
`BigRational` already implements `Iterable<BigRational>`. There are several `set..`. methods to choose an appropriate Iterator, e.g. `setNoDuplicatesIterator()`. 

The power series classes are included in package `edu.jas.ps`, e.g. `UnivPowerSeries` and ...`Ring`. The` order()` method is defined differently, `truncate()` is eventually what is meant by the `order` parameter above.

`UnivPowerSeriesRing` has already a method `seriesOfTaylor()` to compute the Taylor series from a function which has derivatives.

**Comment by kredel** [comment]
Multivariate power series are available in `MultiVarPowerSeries` etc.

**Comment by axkr** [comment]
Ok, is there a way to create a rational function

```java
      BigRational br = new BigRational(0);
      GenPolynomialRing<BigRational> pr = fac.polyRing();

      GenPolynomial<BigRational> numerator = pr.random(...);
      GenPolynomial<BigRational> denominator = pr.random(...);
//???
      TaylorFunction<BigRational> F = new PolynomialTaylorFunction<BigRational>(numerator, denominator);

      UnivPowerSeries<BigRational> ps = fac.seriesOfTaylor(F, br);
```

**Comment by kredel** [comment]
Instead of `PolynomialTaylorFunction​(GenPolynomial<C> p)` it would be
`QuotientTaylorFunction​(GenPolynomial<C> p, q)`. You should implement it similar to PolynomialTailorFunktion, i.e. `derivative()` and `evaluate()` for the univariate case (the multivariate cases are eventually not required).  Using (p/q)' = (p' q + p q')/q**2 (please check).

**Comment by kredel** [comment]
It is minus: (p/q)' = (p' q - p q')/q**2

**Comment by axkr** [comment]
This attempt seems to work (see code below)

I think the `BigRational#multiply()` performance could be improved because very often the GCD will be `1`.

Other implementations move the GCD test into the `BigFraction` constructor to do the "Rational number reduction to lowest terms" only in one place:
- https://github.com/Hipparchus-Math/hipparchus/blob/505e0634f1c02404f5bc31f6de18ff63fe39c22f/hipparchus-core/src/main/java/org/hipparchus/fraction/BigFraction.java#L962

I think method name `deriviative` is a typo you probably mean `derivative` ?

```java
/*
 * $Id$
 */
package edu.jas.ps;


import java.util.List;
import edu.jas.poly.ExpVector;
import edu.jas.poly.GenPolynomial;
import edu.jas.poly.PolyUtil;
import edu.jas.structure.RingElem;

public class PolynomialQuotientTaylorFunction<C extends RingElem<C>> implements TaylorFunction<C> {


  final GenPolynomial<C> numerator;

  final GenPolynomial<C> denominator;

    final long facul;


    public PolynomialQuotientTaylorFunction(GenPolynomial<C> p, GenPolynomial<C> q) {
      this(p, q, 1L);
    }


    public PolynomialQuotientTaylorFunction(GenPolynomial<C> p, GenPolynomial<C> q, long f) {
      this.numerator = p;
      this.denominator = q;
      this.facul = f;
    }


    /**
     * To String.
     * @return string representation of this.
     */
    @Override
    public String toString() {
      return numerator.toString();
    }


    /**
     * Get the factorial coefficient.
     * @return factorial coefficient.
     */
    @Override
    public long getFacul() {
        return facul;
    }


    /**
     * Test if this is zero.
     * @return true if this is 0, else false.
     */
    @Override
    public boolean isZERO() {
      return numerator.isZERO();
    }


    /**
     * Deriviative.
     * @return deriviative of this.
     */
    @Override
    public TaylorFunction<C> deriviative() {
      // (p/q)' = (p' q - p q')/q**2
      GenPolynomial<C> numerDerivative = PolyUtil.<C>baseDeriviative(numerator);
      GenPolynomial<C> denomDerivative = PolyUtil.<C>baseDeriviative(denominator);
      GenPolynomial<C> newNumer =
          denominator.multiply(numerDerivative).subtract(numerator.multiply(denomDerivative));
      GenPolynomial<C> newDenom = denominator.power(2);
      return new PolynomialQuotientTaylorFunction<C>(newNumer, newDenom);
    }
 
    @Override
    public TaylorFunction<C> deriviative(ExpVector i) {
      throw new UnsupportedOperationException();
    }


    /**
     * Evaluate.
     * @param a element.
     * @return this(a).
     */
    @Override
    public C evaluate(C a) {
      C numeratorValue = PolyUtil.<C>evaluateMain(numerator.ring.coFac, numerator, a);
      C denominatorValue = PolyUtil.<C>evaluateMain(numerator.ring.coFac, denominator, a);
      return numeratorValue.divide(denominatorValue);
    }
 
    @Override
    public C evaluate(List<C> a) {
      throw new UnsupportedOperationException();
    }

}
```

**Comment by kredel** [comment]
Looks good. Deriviative is a recuring miss spelling by me. Derivative is correct.

I wonder if it would be meaningful to remove a gcd from `newNumer`, `newDenom`.

`BigRational` is somewhat confusing as `BigInteger` comes from `java.math` not from `edu.jas.arith`, so the respective constructor assumes gcd == 1.

**Comment by axkr** [comment]
> `BigRational` is somewhat confusing as `BigInteger` comes from `java.math` not from `edu.jas.arith`, so the respective constructor assumes gcd == 1.

Yes but why not calculating `BigRational r = RNRED(nu, de);` in a similar fashion for `java.math.BigInteger`?
And using a private `valueOf` method for the case, if you don't need to reduce?

**Comment by axkr** [comment]
A much faster approach seems to be something like this:

```java
      GenPolynomial<BigRational> numerator = ....;
      GenPolynomial<BigRational> denominator = ...;

      final UnivPowerSeries<BigRational> ps;
      BigRational cfac = BigRational.ONE;
      UnivPowerSeriesRing<BigRational> fac = new UnivPowerSeriesRing<BigRational>(cfac);
      TaylorFunction<BigRational> FN = new PolynomialTaylorFunction<BigRational>(numerator);
      if (exprNumerator.isOne()) {
        TaylorFunction<BigRational> FD = new PolynomialTaylorFunction<BigRational>(denominator);
        UnivPowerSeries<BigRational> psD = fac.seriesOfTaylor(FD, BigRational.ZERO);
        ps = psD.inverse();
      } else {
        if (exprDenominator.isOne()) {
          ps = fac.seriesOfTaylor(FN, BigRational.ZERO);
        } else {
          TaylorFunction<BigRational> FD = new PolynomialTaylorFunction<BigRational>(denominator);
          UnivPowerSeries<BigRational> psN = fac.seriesOfTaylor(FN, BigRational.ZERO);
          UnivPowerSeries<BigRational> psD = fac.seriesOfTaylor(FD, BigRational.ZERO);
          ps = psN.divide(psD);
        }
      }
```

**Comment by kredel** [comment]
For the "much faster" post above: it seems that the computing time will depend on the size (=truncate) of the power series. The polynomial degrees are bounded, but the power series eventually not.

There was already a `derivate()` method in `edu.jas.integrate.ElementaryIntegration` - I did not remember - funny.

**Comment by kredel** [comment]
An example and tests have been released with [JAS version 2.7.140](https://github.com/kredel/java-algebra-system/releases/tag/v2.7.140).
